atc/engine/engine.go

```diff
@@ -215,36 +215,38 @@ func (b *engineBuild) Run(ctx context.Context) {
 		}
 	}()
 
-	done := make(chan error)
+	succeeded := false
+
+	done := make(chan struct{})
 	go func() {
+		defer close(done)
 		defer func() {
 			if r := recover(); r != nil {
 				err = fmt.Errorf("panic in engine build step run %d: %v", b.build.ID(), r)
 
 				fmt.Fprintf(os.Stderr, "%s\n %s\n", err.Error(), string(debug.Stack()))
 				logger.Error("panic-in-engine-build-step-run", err)
-
-				done <- err
 			}
 		}()
 
 		ctx := lagerctx.NewContext(ctx, logger)
 		ctx = policy.RecordTeamAndPipeline(ctx, b.build.TeamName(), b.build.PipelineName())
-		done <- step.Run(ctx, state)
+
+		succeeded, err = step.Run(ctx, state)
 	}()
 
 	select {
 	case <-b.release:
 		logger.Info("releasing")
 
-	case err = <-done:
-		logger.Debug("engine-build-done")
+	case <-done:
 		if err != nil {
 			if ok := errors.As(err, &exec.Retriable{}); ok {
 				return
 			}
 		}
-		b.finish(logger.Session("finish"), err, step.Succeeded())
+
+		b.finish(logger.Session("finish"), err, succeeded)
```

Comment:
Is the idea that we don't want to risk setting `succeeded` to a value other than `false`? Technically that would make this part more airtight, but IMO it should be an invariant that when err != nil all other return values are zero-value. Given that assumption, I wouldn't expect to see a test for when a step returns `true, <some err>` so I wouldn't write the code to handle it either.

For that invariant to hold true I guess I would argue that all step tests should assert that the return value is false whenever an error is returned, rather than having tests in the API consumers cover an 'true, error' combination.

Does that make sense?

Code style wise it might make a bit more sense if I had done `var succeeded bool` instead of `succeeded := false` - the direct assignment here is just meant to set the value in the outer scope, which might be more obvious if it's not given an explicit value before. Kind of like this pattern:

```go
var succeeded bool
var err err
if runStep1 {
  succeeded, err = step1.Run()
} else {
  succeeded, err = step2.Run()
}
```

...but in this case `err` is already declared. Maybe I could give it its own name for clarity. 🤔 

```go
var succeeded bool
var stepErr error

done := make(chan struct{})
go func() {
  defer close(done)

  succeeded, stepErr = state.Run(...)
}()
```

Reply:
Thanks. And looking at `b.finish`, it does handle the cases as well.